src/Std/Internal/Http/Data/Header/Name.lean

```diff
@@ -0,0 +1,232 @@
+
+/--
+Creates a new `HeaderValue` from a string with an optional proof of validity.
+If no proof is provided, it attempts to prove validity automatically.
+-/
+@[expose]
+def new (s : String) (h : s.toList.all isValidHeaderCharNode := by decide) : HeaderValue :=
+  ⟨s, h⟩
+
+/--
+Attempts to create a `HeaderValue` from a `String`, returning `none` if the string
+contains invalid characters for HTTP header values.
+-/
+@[expose]
+def ofString? (s : String) : Option HeaderValue :=
+  if h : s.toList.all isValidHeaderCharNode then
+    some ⟨s, h⟩
+  else
+    none
+
+/--
+Creates a `HeaderValue` from a string, panicking with an error message if the
+string contains invalid characters for HTTP header values.
+-/
+@[expose]
+def ofString! (s : String) : HeaderValue :=
+  if h : s.toList.all isValidHeaderCharNode then
+    ⟨s, h⟩
+  else
+    panic! s!"invalid header value: {s.quote}"
+
+/--
+Performs a case-insensitive comparison between a `HeaderValue` and a `String`.
+Returns `true` if they match.
+-/
+@[expose]
+def is (s : HeaderValue) (h : String) : Bool :=
+  s.value.toLower == h.toLower
+
+/--
+Concatenates two `HeaderValue` instances, preserving the validity guarantee.
+-/
+def append (l : HeaderValue) (r : HeaderValue) : HeaderValue :=
+  ⟨l.value ++ r.value, ?_⟩
+where finally
+  unfold isValidHeaderValue
+  rw [String.toList_append, List.all_append, Bool.and_eq_true]
+  constructor
+  · exact l.validHeaderValue
+  · exact r.validHeaderValue
+
+instance : HAppend HeaderValue HeaderValue HeaderValue where
+  hAppend := HeaderValue.append
+
+instance : ToString HeaderValue where
+  toString v := v.value
+
+/--
+Joins an array of `HeaderValue` instances with comma-space separation.
+Returns a single `HeaderValue` containing all values joined together.
+If the array is empty, returns an empty `HeaderValue`.
+-/
+def joinCommaSep (x : Array HeaderValue) : HeaderValue :=
+  if h : 0 < x.size then
+    let first := x[0]'h
+    let rest := x[1...*]
+    rest.foldl (· ++ HeaderValue.new ", " ++ ·) first
+  else
+    .new ""
+
+end HeaderValue
+
+/--
+Checks if a character is valid for use in an HTTP header name.
+-/
+@[expose]
+def isValidHeaderNameChar (c : Char) : Bool :=
+  let v := c.val
+
+  if v < 0x21 || v > 0x7E then
+    false
+  else
+    v != 0x22 && v != 0x28 && v != 0x29 && v != 0x2C && v != 0x3B
+    && v != 0x5B && v != 0x5D && v != 0x7B && v != 0x7D
+
+/--
+Proposition that asserts all characters in a string are valid for HTTP header names.
+-/
+@[expose]
+abbrev isValidHeaderName (s : String) : Prop :=
+  s.toList.all isValidHeaderNameChar ∧ !s.toList.isEmpty
+
+/--
+A validated HTTP header name that ensures all characters conform to HTTP standards.
+Header names are case-insensitive according to HTTP specifications.
+-/
+structure HeaderName where
+  /--
+  The original case-preserved string
+  -/
+  value : String
+
+  /--
+  The proof that it's a valid header name
+  -/
+  validHeaderName : isValidHeaderName value
```

Comment:
I would suggest to have `HeaderName` have an additionnal proof, that the Name is in `normal` form or `canonical` form. I think `normal` is easier.
This will allow to set the internal representation to Lowercase only once, and evoid some computation for each `hash`. The custom `Hash` instance can then also be removed.

```lean4
@[expose]
abbrev isNormalform (s: String) : Prop := s = s.toLower
```
then

```suggestion
structure HeaderName where
  /--
  The original case-preserved string
  -/
  value : String

  /--
  The proof that it's a valid header name
  -/
  validHeaderName : isValidHeaderName value
  /--
  The proof that we stored the header name in normal form
  -/
  normalform: isNormalForm value
```